Implement unixfs sharding

whyrusleeping opened the pull request:
This PR implements a hash array mapped trie using merkledag nodes as a base layer for sharded unixfs directories. 

This changeset is broken up into three distinct parts:
1. Implement the HAMT structure itself. This is a bunch of algorithmic code that i'd like some well thought out sanity checking on.
   - Need confirmation on hash code numbering
   - Thoughts on willf/bitset vs stdlib `math/big.Int`
   - Implementation correctness
   - Consensus on addition of fields to unixfs protobuf
2. Integration of HAMT code into the directory builder logic.
   - provides us a nice abstraction to think about things in terms of a filesystem
   - optimizes smaller directories by not sharding until we hit a threshold
3. Integration of directory builder code into mfs
   - moves mfs from using a simple merkledag `Node` object to using a dirbuilder
   - allows mfs to create sharded directories
   - by way of the previous point, `ipfs add` and friends can now create sharded directories.

Thorough review will be very appreciated!

cc @Kubuxu @jbenet @kevina @diasdavid @dignifiedquire 

also cc @haadcode as you might like to review the datastructure logic

whyrusleeping commented:
Note: tests won't pass yet as I havent vendored some of the new libs. I havent quite decided between [willf/bitset](https://github.com/willf/bitset) and just using the `math/big.Int` to handle the bitfield logic.

whyrusleeping commented:
going with the stdlib, fewer imports and the stdlib is proven to be pretty stable. Its a hair bit less efficient on some operations, but we can optimize later if it becomes an issue.

kevina commented:
@whyrusleeping, I am afraid I do not know anything about sharding, so I will have to first learn about that before I am of any use.  Doing a quick search I discovered ipfs/notes#76, is there anything else that would help with content for this pull request?

jbenet commented on unixfs/hamt/hamt.go:
```diff
@@ -0,0 +1,393 @@
+)
+
+const (
+	HashMurmur3 uint64 = iota
```
remember that this is configurable on the object

jbenet replied:
it will likely use multihash if this proposal goes through: https://github.com/multiformats/multihash/issues/30

whyrusleeping replied:
Yeah, i'd love to use an 'official' multihash table for this.

jbenet commented:
@whyrusleeping this is a very big PR that will be critical to review + merge in stages. if this has any bugs it will be serious trouble and will take weeks of work to find and debug. please:

- rip out unixfs into its own repo first, then we can iterate on the datastructure there first. once we're convinced it really works, we can pull this back into ipfs.
- make it easy to test and verify the hamt data struct stuff is working well. meaning make some tests that shard large directories well. and be able to produce graphviz out of these to inspect and debug formation. 
- this is going to need very serious stress tests. so far i saw very light invariance testing but no serious stresses, with long sequences of operations expected to test how the HAMT grows and shrinks as stuff is added, or testing that the exact same data structures are produced despite sequences of operations being different.

whyrusleeping commented:
@kevina if you want to review the algorithm for the HAMT datastructure and use that to check my implementation, that would be very helpful.

https://en.wikipedia.org/wiki/Hash_array_mapped_trie

whyrusleeping commented:
@jbenet I moved the HAMT code into its own repo and added a larger test to verify operation order doesnt change the resultant structure: https://github.com/whyrusleeping/unixfs-hamt

the `hamt_stress_test.go` will generate a random set of insertions and deletions and run through them in different permutations and verify the outputs are consistent.

jbenet commented:
Thanks i will revivew this week

diasdavid commented:
@whyrusleeping @jbenet mentioned that it would be better rip out all the unixfs into its own repo and this actually might be extremely beneficial to point several people into something isolated they can review fully.

Also, I think the best review I can do is grab this implementation and write the implementation in JavaScript and possibly write the spec for it.

whyrusleeping commented:
@diasdavid I had it pulled out at one point, but it was really just for the sake of 'putting the code somewhere else to look at it'. Actually extracting this code is difficult because it would require extracting a large number of other go-ipfs packages.

whyrusleeping commented:
Rebased on latest master (1a365a86fe2048705052ed1feb33cccbf893eb9b).

@diasdavid @dignifiedquire The actual HAMT code is isolated in this single file: https://github.com/ipfs/go-ipfs/blob/07683a7d9a6e83c859c84e75b2eb47d2aee79aa7/unixfs/hamt/hamt.go

All the review i'm looking for is just that file. the rest of the changes are just integration of that up through the stack and i'm not nearly as concerned about that part of the changeset.

dignifiedquire commented on unixfs/hamt/hamt.go:
```diff
@@ -0,0 +1,466 @@
+	mask := new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(int64(bp)), nil), big.NewInt(1))
+	mask.And(mask, ds.bitfield)
+
+	return popCount(mask)
```
I'm not entirely clear on how this function works/what exactly it's purpose is. And in turn I'm having a hard time verifying that the bit operations you do, do what they are supposed to be doing. Would be great to have a comment about what the function does and an explenation why the bit ops do what they do.

dignifiedquire replied:
Not the `popCount` that one is clear, but rather the whole `indexForBitPos`.

dignifiedquire commented on unixfs/hamt/hamt.go:
```diff
@@ -0,0 +1,466 @@
+package hamt
+
+import (
```
To make it easier for people reading the code in the future it would be great to have a comment here at the top describing

- the exact type of HAMT that is implemented
- specific decisions made for this implementation

dignifiedquire commented on unixfs/hamt/util.go:
```diff
@@ -0,0 +1,61 @@
+	"math/big"
+)
+
+type hashBits struct {
```
Would be nice to have some comments explaining what exactly the purpose of `hashBits` as a data structure is and how `.Next` works/what it returns

dignifiedquire commented:
Reviewed the test cases + the actual implementation in `hamt.go` from a algorithmic perspective, not from a go-code perspective. As far as I can tell this looks good to me 👍 Thanks @whyrusleeping for the additional comments, the make it a lot easier to follow the code along now.

Kubuxu commented:
One issue I can see is that we never downgrade form HMAT. This means that of someone reaches threshold and then removes files below treshold. We could downgrade but it could end up expensive when someone hovers (add del add del) around the threshold.

whyrusleeping commented:
@Kubuxu Yeah, you can manually downgrade by simply creating a new directory and copying each child into the new directory (as long as its below the sharding limit). We can add tooling for that later to help with that if it becomes an issue.

Kubuxu commented:
I was writing it on a phone in a bus and I lost part:
This means that if someone reaches threshold and then removes files below threshold he will get different hash which might be something that he is not expecting.

Also the value of threshold is now arbitrary, should we reason and try to find "best" threshold?

whyrusleeping commented:
@Kubuxu 

> Also the value of threshold is now arbitrary, should we reason and try to find "best" threshold?

Yeah, It needs to be small enough that moving from unsharded to sharded isnt really expensive, but large enough that its not wasteful to be using a hamt at that layer.

I'm open to suggestions

Kubuxu commented on unixfs/hamt/hamt.go:
```diff
@@ -0,0 +1,491 @@
+)
+
+const (
+	HashMurmur3 uint64 = 0x22
```
Should we add murmur to multicodec/multihash?

whyrusleeping replied:
We have a multicodec for it IIRC...

Oh, looks like @diasdavid ignored my PR and changed a bunch of stuff... :/

It was here: https://github.com/multiformats/multicodec/pull/23/files

diasdavid replied:
I didn't ignore @whyrusleeping, I had asked a question on it -- https://github.com/multiformats/multicodec/pull/23#discussion_r86863096 --, the only thing that happened was that the main one got merged, so now you can PR to master :) 

Btw, new feature on github, if you click "edit" at the top level of the PR, you can pick a new target:

![image](https://cloud.githubusercontent.com/assets/1211152/20568876/d72fef48-b195-11e6-96b1-47a283d95e13.png)

You don't have to close and open again

Kubuxu commented on unixfs/hamt/hamt.go:
```diff
@@ -0,0 +1,491 @@
+	Label() string
+}
+
+func NewHamtShard(dserv dag.DAGService, size int) *HamtShard {
```
Probably makeHamtShard function should enforce it.

lgierth commented on test/sharness/t0250-files-api.sh:
```diff
@@ -481,7 +508,7 @@ test_files_api() {
 
 # test offline and online
-test_files_api
+#test_files_api
```
What about this disabled test?

Kubuxu replied:
Same

whyrusleeping replied:
oh, huh. k

lgierth commented:
This looks good to me overall

todo:
- [ ] disabled test (see other comment)
- [ ] murmur3 multicodec
- [ ] decide on downgrading -- I personally think it's pretty important

jbenet commented:
**Important!** If you make a threshold that isn't _perfectly deterministic_, you violate the entire point of using a CHAMP HAMT in the first place. Meaning: **don't do that!**. 

Either:
- [ ] switch to HAMT entirely (dont use the old data structure at all, no threshold madness)
- [ ] make it _perfectly deterministic_.  Trees with given children should ALWAYS be the exact same root hash, no matter what order of operations was involved.

Important requirement before merging:
- [ ] Make a test case that VERIFIES this stability. We cannot bear any non-stable data structure here.

whyrusleeping commented:
@lgierth @jbenet what should happen if someone tries modifying an old style directory?

kevina commented:
Could we make the new style directory a distinct Cid type?  Just a possibility.

whyrusleeping commented:
@kevina eh... not really. Then it wouldnt be unixfs anymore. Plus, we can already tell if its a shard or not, we have the unixfs type field.

kevina commented:
Okay, just wondering.  It might have solved the problem, if its the old CId type than it is not HAMT if it is the new than it is either the old or HAMT based on the content (likely the size or number of entries.

whyrusleeping commented:
@kevina thats not the problem we're worried about. The issue is that given any method of creating a directory (any number of insertions or deletions in any order), we should always get the same exact object for the same set of entries.

the issue raised above is that if i add 1001 files to a directory, and then delete two, it won't be the same output object as if i had just added 999 in the first place.

kevina commented:
@whyrusleeping I was trying to help solve:

> what should happen if someone tries modifying an old style directory?

If not a CId type we could have one a new unixfs type field one for the old (say 0x1), one for the old that could be converted to a HAMT (0x2) is the numbers of files is over the threshold (say 1000) (0x3), and another for the new HAMT type.  If the unixfs type is 0x1 than that will never become 0x3, it it is 0x2 then if the size is over 1000 it will become 0x3.  Having a 0x2 type with a size of over 1000 would be disallowed.  In a similar manor having a size less than 1000 and a type of 0x3 will be disallowed.

The flipflopping between 0x2 and 0x3 will be unavoidable if you want to be deterministic.  If the user knows the size is going to be around this threshold they can explicitly set it to 0x1.

I hope this makes sense, if I am just adding noise to the conversation I apologize.

dignifiedquire commented:
@whyrusleeping could you rebase this onto `master` please?

whyrusleeping commented:
rebased!

whyrusleeping commented:
@pgte thanks for reporting, found the problem. The filestore node wrappers were getting passed places they werent welcome.

whyrusleeping commented:
Alright. I'm gonna rebase this, then add a flag to disable it so we can ship 0.4.7 with support for this via an experiments flag.

pgte commented:
@whyrusleeping One last disparity from the JS implementation:

Looks like the UnixFS fileSize is not being populated, and probably should.

/cc @diasdavid

dignifiedquire commented:
> Alright. I'm gonna rebase this, then add a flag to disable it so we can ship 0.4.7 with support for this via an experiments flag.

@whyrusleeping maybe in 0.4.8? :)

Kubuxu commented:
Wow, rebased. Is this coming in soon (TM)?

whyrusleeping commented:
@pgte Anything else we need to finish before this goes forward? I think we've agreed on not setting filesizes on the shards.

pgte commented:
@whyrusleeping Correct, I don't think we need anything else.

Kubuxu commented:
Apart from one removed test: https://github.com/ipfs/go-ipfs/pull/3042#discussion_r107789873

Kubuxu commented on mfs/dir.go:
```diff
@@ -235,9 +235,9 @@ func (d *Directory) ListNames() ([]string, error) {
-func (d *Directory) List() ([]NodeListing, error) {
+func (d *Directory) List(ctx context.Context) ([]NodeListing, error) {
 	var out []NodeListing
-	err := d.ForEachEntry(context.TODO(), func(nl NodeListing) error {
```
I shouldn't have missed that.

whyrusleeping commented:
Choo Choo! Here comes the merge train! :train::train::train::train::train::train::train::train:
